File: 2020/11/week2/jjeda/MatchStick.java
public class MatchStick {

  private static int[] cost = { 6, 2, 5, 5, 4, 5, 6, 3, 7, 6 };
  private static long count;

  private long solution(int k) {
    count = 0;
    process(k, 0, -1);
    return count;
  }

  private void process(int num, int depth, int before) {
    if (num == 0) {
      count++;
      return;
    }
    if (before == 0 && depth == 1) {
      return;
    }

    for (int i = 0; i < 10; i++) {
      if (num - cost[i] >= 0) {
        process(num - cost[i], depth + 1, i);
      }
    }
  }
}

File: 2020/11/week2/jjeda/MatchStick.kt
class MatchStick {
  private val cost = intArrayOf(6, 2, 5, 5, 4, 5, 6, 3, 7, 6)
  private val cache = Array(51) { 0 }

  fun solution(k: Int): Int {
    cost.forEach { requiredMatches ->
      cache[requiredMatches]++
    }
    cache[6]--

    for (i in 4..k) {
      cost.filter { it < i }
          .forEach { cache[i] += cache[i - it] }
    }
    cache[6]++
    return cache[k]
  }
}